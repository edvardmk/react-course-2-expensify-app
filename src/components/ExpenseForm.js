import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/initialize';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      description: props.expense?.description || '',
      note: props.expense?.note || '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: moment(props.expense?.createdAt),
      calendarFocused: false,
      error: ''
    }
  }

  onDescriptionChange = (e) => {
    this.setState({ description: e.target.value })
  }
  onNoteChange = (e) => {
    this.setState({ note: e.target.value })
  }
  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount })
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState({ createdAt })
    }
  }
  onFocusedChange = ({ focused }) => {
    this.setState({ calendarFocused: focused })
  }
  onSubmit = (e) => {
    e.preventDefault()

    if (!this.state.description || !this.state.amount) {
      this.setState({ error: 'Please provide description and amount.' })
    } else {
      this.setState({ error: '' })
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      })
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
           <input 
             type="text"
             placeholder="Description"
             autoFocus
             value={this.state.description}
             onChange={this.onDescriptionChange}
           />
           <input
             type="text"
             placeholder="Amount"
             value={this.state.amount}
             onChange={this.onAmountChange}
           />
           <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusedChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
           />
           <textarea
            placeholder="Add a note to your Expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
           >
           </textarea>
           <button>
            Add Expense
           </button>
        </form>
      </div>
    )
  }
}