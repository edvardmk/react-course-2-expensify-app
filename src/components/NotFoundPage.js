import React from 'react'
import { Link } from 'react-router-dom'

const NotFoundPage = () => (
  <Link to="/">404!</Link>
)

export default NotFoundPage