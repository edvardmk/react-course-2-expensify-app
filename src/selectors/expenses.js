import moment from 'moment'

export default (expenses, { startDate, endDate, text, sortBy }) => {
  return expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
    const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return b.createdAt - a.createdAt
    }
    if (sortBy === 'amount') {
      return b.amount - a.amount
    }
  })
}
