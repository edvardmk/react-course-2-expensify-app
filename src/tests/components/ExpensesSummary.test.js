import React from 'react'
import { shallow } from 'enzyme'

import { ExpensesSummary } from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'

test('should render ExpensesSummary with one expense correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={123} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary with multiple expenses correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={3} expensesTotal={1234} />)
  expect(wrapper).toMatchSnapshot()
})