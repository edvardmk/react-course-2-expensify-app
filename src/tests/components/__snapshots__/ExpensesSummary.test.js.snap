// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with multiple expenses correctly 1`] = `
<p>
  Viewing 
  3
   
  expenses
   totalling 
  $12.34
</p>
`;

exports[`should render ExpensesSummary with one expense correctly 1`] = `
<p>
  Viewing 
  1
   
  expense
   totalling 
  $1.23
</p>
`;
